### PR TITLE
fix: add comprehensive S3 SDK test coverage from e2e suite

### PR DESCRIPTION
**Pull request by ovaistariq**
## Summary

Add extensive SDK test coverage for S3 operations adapted from the Tigris e2e test suite. Tests validate proxy behavior end-to-end including PutObject (1MB), multipart uploads (small parts, multiple parts, abort flow), range requests (11 edge cases), content encoding preservation, and checksum verification (SHA256, CRC32, CRC32C).

## Changes

- **tests/s3compat/sdk/go_sdk_test.go**: Added 8 new test functions covering PutObject at 1MB scale, multipart upload scenarios, comprehensive range request handling, content-encoding preservation, and checksum verification for multiple algorithms
- **s3err/errors.go**: Extracted error definitions to new package (refactoring from handlers/errors.go)
- **handlers/errors.go**: Consolidated logging and removed debug output
- **proxy modules**: Various minor improvements to logging and error handling

## Test plan

All new tests are table-driven subtests running against external Tigris instances:
- `TestSDK_PutObject_1MB` - validates 1MB PUT/GET round-trip
- `TestSDK_MultipartUpload_MultipleParts` - two 5MB parts with full content verification
- `TestSDK_MultipartUpload_SmallPart` - single 16KB part edge case
- `TestSDK_MultipartUpload_Abort` - abort flow verification
- `TestSDK_GetObjectRange_Comprehensive` - 11 range scenarios with byte-level verification
- `TestSDK_GetObjectRange_Invalid` - error handling for invalid ranges
- `TestSDK_ContentEncodingPreservation` - 512B and 1MB objects with gzip encoding
- `TestSDK_PutObject_Checksum` - 6 subtests (SHA256/CRC32/CRC32C × 512B/32KB)
- `TestSDK_MultipartUpload_Checksum` - SHA256 checksum through multipart flow

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Behavior changes affect HTTP header forwarding/caching and S3 error responses, which can impact client compatibility; the large new e2e-style tests help reduce regression risk.
> 
> **Overview**
> Adds extensive new S3 SDK compatibility tests (1MB PUT/GET, multiple multipart flows including abort, comprehensive/invalid range GETs, Content-Encoding preservation, and checksum propagation for single and multipart uploads).
> 
> Improves proxy/cache correctness by caching and replaying additional object response headers (e.g., `Content-Encoding`, disposition/language/expiry, SSE, versioning, multipart parts count), adding `WriteHeaders` options for range responses, and ensuring cached range requests return proper `206` headers and `InvalidRange` errors.
> 
> Refactors S3 XML error definitions/writers into a new `s3err` package and updates handlers/proxy to use it, while also preventing Go’s HTTP client from auto-decompressing upstream responses (`DisableCompression: true`) and adding more targeted debug logging (upstream latency, authz revocations, and only logging newly learned signing keys).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fa9ca9483d130dc8317b7ddf2478c6f574a57e6f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by efirs — APPROVED** [review]

**Comment by ovaistariq** [comment]
@greptileai

File: proxy/get_object.go [hunk 1]
```diff
@@ -587,21 +589,15 @@ func (s *Service) serveRangeFromCache(
 	// Parse Range header
 	ranges, err := parseRangeHeader(rangeHeader, meta.ContentLength)
 	if err != nil || len(ranges) == 0 {
-		// Invalid range - return 416 Range Not Satisfiable
-		w.Header().Set("Content-Range", fmt.Sprintf("bytes */%d", meta.ContentLength))
-		w.Header().Set(XCacheHeader, XCacheHit)
-		w.WriteHeader(http.StatusRequestedRangeNotSatisfiable)
+		s3err.WriteError(w, r, s3err.ErrInvalidRange)
```
[the remark this answers is not in the document]
Comment (ovaistariq) [reply proxy/get_object.go:594]: Fixed

File: proxy/get_object.go [hunk 2]
```diff
@@ -587,21 +589,15 @@ func (s *Service) serveRangeFromCache(
...
 	}
 
 	// Only support single range (multi-range is complex and rare)
 	if len(ranges) > 1 {
 		log.Debug().Str("bucket", bucket).Str("key", key).Msg("Multi-range not supported from cache")
-		// For multi-range, return 416 - we don't support it
-		w.Header().Set("Content-Range", fmt.Sprintf("bytes */%d", meta.ContentLength))
-		w.Header().Set(XCacheHeader, XCacheHit)
-		w.WriteHeader(http.StatusRequestedRangeNotSatisfiable)
+		s3err.WriteError(w, r, s3err.ErrInvalidRange)
```
[the remark this answers is not in the document]
Comment (ovaistariq) [reply proxy/get_object.go:604]: Fixed

**Comment by ovaistariq** [comment]
@greptileai

**Comment by ovaistariq** [comment]
@greptileai
